### PR TITLE
Initial implementation of Picker

### PR DESCRIPTION
**Pull request by devongovett**
This is an initial version of the `Picker` component, along with the `useSelect` aria hook and `useSelectState` stately hook. See individual commit messages for some details about related changes.

A few notable new features:

* You can type to select when focus is on the field button, without opening the menu. This is similar to how a native `<select>` works.
* The menu opens on mouse down rather than on mouse up, and you can drag over an item with your mouse still down and release to select that item. Also similar to how a native `<select>` works.

I also had to refactor ListBox into ListBoxBase so that Picker can access it directly. This is to allow the collection state and the layout to be held in the Picker rather than the ListBox. This enables type to select without opening the menu, along with layout info to be cached rather than recomputed every time the ListBox is mounted.

There is still a bunch left to do, including tests and a few more features and bug fixes. Just wanted to get an initial implementation in first and iterate on that rather than a giant PR.

### CODE REVIEW

File: packages/@react-aria/interactions/src/usePress.ts [hunk 1]
```diff
@@ -151,6 +152,23 @@ export function usePress(props: PressHookProps): PressResult {
...
+    let triggerPressUp = (originalEvent: EventBase, pointerType: PointerType) => {
+      if (isDisabled) {
+        return;
+      }
+
+      if (onPressUp) {
+        onPressUp({
+          type: 'pressup',
+          pointerType,
+          target: originalEvent.target as HTMLElement,
```
Comment (snowystinger) [inline packages/@react-aria/interactions/src/usePress.ts]: spread the original event?

File: packages/@react-aria/menu/src/useMenuTrigger.ts [hunk 2]
```diff
@@ -70,7 +70,7 @@ export function useMenuTrigger(props: MenuTriggerProps, state: MenuTriggerState)
     menuTriggerProps: {
       ...triggerAriaProps,
       id: menuTriggerId,
-      onPress,
+      onPressStart: onPress,
```
Comment (snowystinger) [inline packages/@react-aria/menu/src/useMenuTrigger.ts]: I don't follow this change, can you explain?

File: packages/@react-aria/overlays/src/useOverlayTrigger.ts [hunk 3]
```diff
@@ -53,14 +53,21 @@ export function useOverlayTrigger(props: OverlayTriggerProps): OverlayTriggerAri
     };
   }, [isOpen, onClose, ref]);
 
+  // Aria 1.1 supports multiple values for aria-haspopup other than just menus.
+  // https://www.w3.org/TR/wai-aria-1.1/#aria-haspopup
+  // However, we only add it for menus for now because screen readers often
```
Comment (snowystinger) [inline packages/@react-aria/overlays/src/useOverlayTrigger.ts]: this comment is now untrue?

File: packages/@react-aria/selection/src/useTypeSelect.ts [hunk 4]
```diff
@@ -41,7 +49,19 @@ export function useTypeSelect(options: TypeSelectOptions): TypeSelectAria {
...
+function getStringForKey(key: string) {
+  // If the key is of length 1, it is an ASCII value.
+  // Otherwise, if there are no ASCII characters in the key name,
+  // it is a Unicode character.
+  // See https://www.w3.org/TR/uievents-key/
+  if (key.length === 1 || !/^[A-Z]/i.test(key)) {
+    return key;
+  }
+
+  return '';
```
Comment (snowystinger) [inline packages/@react-aria/selection/src/useTypeSelect.ts]: will this prevent languages like Chinese?

File: packages/@react-stately/collections/src/ListLayout.ts [hunk 5]
```diff
@@ -57,6 +57,8 @@ export class ListLayout<T> extends Layout<Node<T>> implements KeyboardDelegate {
   private indentationForItem?: (collection: Collection<Node<T>>, key: Key) => number;
   private layoutInfos: {[key: string]: LayoutInfo};
   private contentHeight: number;
+  collection: Collection<Node<T>>;
```
Comment (snowystinger) [inline packages/@react-stately/collections/src/ListLayout.ts]: private?

File: packages/@react-types/shared/src/events.d.ts [hunk 6]
```diff
@@ -30,7 +30,7 @@ export type FocusEvent = BaseEvent<ReactFocusEvent<any>>;
 export type PointerType = 'mouse' | 'pen' | 'touch' | 'keyboard';
 
 export interface PressEvent {
-  type: 'pressstart' | 'pressend' | 'press',
+  type: 'pressstart' | 'pressend' | 'pressup' | 'press',
```
Comment (snowystinger) [inline packages/@react-types/shared/src/events.d.ts]: what's the difference between pressup and pressend?

File: packages/@react-aria/menu/src/useMenuTrigger.ts (hunk 2 above)
Comment (devongovett) [reply packages/@react-aria/menu/src/useMenuTrigger.ts]: see PR description.

File: packages/@react-aria/selection/src/useTypeSelect.ts (hunk 4 above)
Comment (devongovett) [reply packages/@react-aria/selection/src/useTypeSelect.ts]: nope, see above comment.

File: packages/@react-aria/overlays/src/useOverlayTrigger.ts (hunk 3 above)
Comment (devongovett) [reply packages/@react-aria/overlays/src/useOverlayTrigger.ts]: menus and listboxes are close enough where if it is exposed as a menu it's not that big of a deal. we did this in v2.

File: packages/@react-stately/collections/src/ListLayout.ts (hunk 5 above)
Comment (devongovett) [reply packages/@react-stately/collections/src/ListLayout.ts]: nope, it must be set by the user

File: packages/@react-types/shared/src/events.d.ts (hunk 6 above)
Comment (devongovett) [reply packages/@react-types/shared/src/events.d.ts]: press up is like mouse up: fired whenever the user releases a pointer over the target. press end is fired whenever the user's pointer leaves the target or is released. press is fired if there is a pointer down event followed by a pointer up event over the target.

File: packages/@react-aria/interactions/src/usePress.ts (hunk 1 above)
Comment (devongovett) [reply packages/@react-aria/interactions/src/usePress.ts]: That would include a lot more properties than we want.

**Pull request merged**